### PR TITLE
Implement kubelet CredentialProvider

### PR DESCRIPTION
**Pull request by tjamet**
Context
---

We currently suport 2 authentication sources. Docker and containerD.

Problems
---

1. Kubelet has a third one: [CredentialProviderConfig]

To be consistent, noe needs to implement all of those to be able to
lookup the values in the relevant place.

2. The current containerD authentication is mixed with the Docker one (file
and image pull secret ones).

Goal
---

Add supoort for kubelet CredentialProvider

[CredentialProviderConfig]: https://kubernetes.io/docs/tasks/administer-cluster/kubelet-credential-provider/
<details>
<summary>
Committer details
</summary>
Local-Branch: main
</details>
<details>
<summary>
Related changes
</summary>
<details>
<summary>
Future changes
</summary>
<details>
<summary>
Improve Readme, detail helm chart values (#120)
</summary>

</details>
</details>
</details>

### CODE REVIEW

File: pkg/registry/kubelet_credentials.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+
+	return s
+}
+
+func parseObjects(scheme *runtime.Scheme, r io.Reader, object runtime.Object) error {
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return err
+	}
+	dfltGVK := (&kubeletcredentialsproviderv1.CredentialProviderResponse{}).GroupVersionKind()
```
Comment (Fsero) [inline pkg/registry/kubelet_credentials.go:69]:
```suggestion
	defaultGVK := (&kubeletcredentialsproviderv1.CredentialProviderResponse{}).GroupVersionKind()
```

maybe more readable

**Review by Fsero — APPROVED** [review]

**Comment by dackroyd** [comment]
Need assistance here? My team is looking to use noe with images sourced from ECR, and this looks preferable to adding ECR specific support

**Comment by Fsero** [comment]
@dackroyd i think this one only needs fixing conflicts to be good to merge, i'd take a look during the week :)

**Comment by dackroyd** [comment]
@Fsero I got this running today in minikube, with the ECR credential provider. Ran into issues there with credentials not being provided due to there being some informational output on stderr from the provider, in addition to the token on stdout. Even if the call is a success, if anything is output to stderr, it’s being considered a failure.

Also ran into some other issues with the helm chart, specifically flags being specified incorrectly so not taking effect.

**Comment by dackroyd** [comment]
Key log entry for the ECR provider is this one:

https://github.com/kubernetes/cloud-provider-aws/blob/master/cmd/ecr-credential-provider/main.go#L129

```
func (e *ecrPlugin) getPrivateCredsData(imageHost string, image string) (*credsData, error) {
	klog.Infof("Getting creds for private image %s", image)
```

Example output for that on `stderr`:

```
I0509 00:14:28.190934      72 main.go:124] Getting creds for private image {image}
```

In `kubelet_credentials.go`:

```
	err = execCommandOutput(ctx, &stdin, &stdout, &stderr, kubeToExec(provider.Env), filepath.Join(r.BinDir, provider.Name), provider.Args...)
	if err != nil {
		log.DefaultLogger.WithContext(ctx).WithError(err).Error("Could not execute kubelet credentials provider, skipping it")
		return
	}
	if stderr.Len() > 0 {
		log.DefaultLogger.WithContext(ctx).WithField("error", "stderr").Warn(stderr.String())
		return
	}
```

^ Anything output to stderr will cause the process to fail

**Comment by tjamet** [comment]
Hi @dackroyd . Thank you for your detailed feedback.
I have changed how we handle stderr outputs so it does not fail any longer.
I would expect to be able to merge this in the coming days after some further testing

File: charts/noe/values.yaml [hunk 2]
```diff
@@ -14,6 +14,17 @@ matchNodeLabels: []
 # - accelerator.node.kubernetes.io/inference
 # - accelerator.node.kubernetes.io/gpu
 
+kubeletConfig:
+#   binDir: /etc/eks/image-credential-provider
+#   configDir: /etc/eks/image-credential-provider
+#   config: config.json
+
+pod:
```
Comment (Fsero) [inline charts/noe/values.yaml:22]: we should bump the chart version this is kind of a braking change

**Review by Fsero — APPROVED** [review]
lgtm

**Review by carlosrejano — APPROVED** [review]

**Pull request merged**